### ISSUE TITLE
freeze all SHAs, move them forward together weekly?

### ISSUE THREAD

**Issue by SethTisue** [scala/community-builds#338]
we track master of many projects now, especially in the 2.12 build

that's **good** because
- we find out about problems quickly

but it's also **bad** because
- anytime something fails, we never know right away whether it's because Scala changed, or because the project changed
- it slows down maintenance of the community build, because you can be in the middle of testing some change, when suddenly some cornerstone project like ScalaTest or Specs2 merges a trivial PR, and now everything has to be rebuilt and the change you thought would take 10 minutes to test suddenly takes 8 hours

**proposed fix**:
- freeze all projects to fixed SHAs
- write a script that bumps all the SHAs to the latest
  - with some mechanism for leaving certain projects frozen, perhaps a special comment on that line
- ...and submits a pull request and starts a test run
- run the script weekly or so
- if the test run is green, merge the PR, if not, investigate

**or maybe**: it would also be possible to have separate "stable" and "bleeding edge" community builds, but I'm not sure maintaining them and keeping them in sync and making the necessary Jenkins changes in Chef is worth the effort. I think the above scheme might strike a nice cost/benefit balance

**Comment by dwijnand** [comment]
I think your proposal is very viable.

But I wonder if perhaps the opposite makes sense - given:
- the projects in community build change much more frequently than scala/scala
- any reasonably aggressive change in scala/scala is tested with the community build even before merging

Why not freeze scala/scala instead, and bump at some interval (nightly/weekly/etc)?

**Comment by adriaanm** [comment]
Here's how I'm implementing this:

To turn the repeatable build json into a list of URIs, extract the JSON between

```
[info] You can repeat this build (except for -SNAPSHOT references) using this configuration:
```

and

```
[info] ---== End Repeatable dbuild Configuration ===---
```

```
curl -s \
https://scala-ci.typesafe.com/view/scala-2.12.0/job/scala-2.12.0-integrate-community-build/17/consoleText \
| awk '/.*You can repeat this build.*using this configuration:/{y=1;next}/.*---== End Repeatable dbuild Configuration ===---/'{y=0}y \
| jq -re '.build.projects[] | "\(.name)-uri: \"\(.uri)\""' | grep -v scala-uri | sort
```

First shot: https://github.com/scala/community-builds/compare/2.12.x...adriaanm:2.12.x?expand=1
Validation:  https://scala-ci.typesafe.com/job/scala-2.12.x-integrate-community-build/796/console

**Comment by milessabin** [comment]
@dwijnand @SethTisue your proposals aren't mutually exclusive ... we could do both.

**Comment by SethTisue** [comment]
in the future, when we freeze all the SHAs, it will be vital to take steps to make sure that the commits we froze at don't disappear from GitHub. this has already happened to four projects from the 2.12.0 freeze, as detailed over at https://github.com/scala/community-builds/issues/401

basically I think we need to make a new GitHub organization, as originally planned, and fork not just certain repos we made custom changes to, but fork all of them, and push tags to all of the forks so the commits we froze them at can never become dangling even when we merge subsequent changes from upstream. (doing this will also protect us an against entire repository disappearing or being moved or renamed)

**Comment by SethTisue** [comment]
@dwijnand I like your suggestion, but we can't _only_ do it that way — in practice, I'm afraid it's all too common for a Scala change to break something. it's already happened twice since 2.12.0. we would need to (as Miles suggests) do both, one where Scala is frozen and the projects move, and another where the projects are frozen and Scala moves.

**Comment by ghost** [comment]
Thought: If it were easier/easy for the community build (CB) to be run locally (ie by project maintainers) they would be able to run their changes against the CB _before_ merging to master. I'm not suggesting they do this as part of, say, a Travis build nor even every merge.

But at times projects might actually want to do this - "I wonder if this will break anything". Just a thought on how to get others on board, helping

**Comment by SethTisue** [comment]
https://github.com/scala/community-builds/pull/492 added Adriaan's freeze script to the repo

**Comment by SethTisue** [comment]
I think the current state of this is good enough. Adriaan's script exists when we need to do a mass freeze. other decisions about whether to freeze or track a branch can be made a repo-by-repo basis. plus we can adjust the overall tendency to track vs. freeze — right now I lean pretty heavily towards track, but I can allow myself more freezing in cases where tracking gets annoying. it's fine.

**Comment by SethTisue** [comment]
Reopening because I've been thinking about this again. It's annoying constantly having projects changing out from under me when I'm working on the community build. For example, the Akka team is very industrious and is always committing stuff, and the akka-actor tests take a while to run, and then the akka-more tests take quiiiiiite a while to run (we weren't running them before, now we are).

Leaning more and more towards thinking "frozen" should be the default state, and moving stuff forward should be a conscious choice.

**Comment by SethTisue** [comment]
> Leaning more and more towards thinking "frozen" should be the default state, and moving stuff forward should be a conscious choice

ok, I've finally flipped the switch on this. all the builds are now fully frozen except when I'm actually working on something.

2.11 I'll unfreeze-then-refreeze every 1-2 months, 2.12 and 2.13 maybe every 2-3 weeks, something like that

if a Scala release is coming up or when some particularly important/risky Scala PR gets merged, I think it's fine to just bump the Scala SHA and not worry about rushing an unfreeze all the project SHAs. we are not substitute CI for the whole community, having older versions of things is fine, it's enough for validating Scala releases.

**Comment by SethTisue** [comment]
> in the future, when we freeze all the SHAs, it will be vital to take steps to make sure that the commits we froze at don't disappear from GitHub

this has turned out to be less of a problem than I had feared. I think at the time I wrote the above, we had just lost a commit, but in the years since, it has happened pretty much never. I think maybe one time I had to rescue a commit from a clone (I don't remember whether I used a clone on one of my laptops, or a clone on one of the Jenkins servers).

it's not worth adding extra infra and extra hassle over, I now think. even if we occasionally end up losing a commit, finding a substitute commit should be no big deal.
